consensus/safety-rules/src/safety_rules.rs

```diff
@@ -230,11 +243,37 @@ impl<T: Payload> TSafetyRules<T> for SafetyRules<T> {
         ))
     }
 
-    /// @TODO only sign blocks that are later than last_voted_round and match the current epoch
-    /// @TODO verify QC correctness
-    /// @TODO verify QC matches preferred round
     fn sign_proposal(&mut self, block_data: BlockData<T>) -> Result<Block<T>, Error> {
         debug!("Incoming proposal to sign.");
+        self.verify_author(block_data.author())?;
+        self.verify_epoch(block_data.epoch())?;
+        let last_voted_round = self.persistent_storage.last_voted_round()?;
```

Comment:
would recommend a new line between 249 and 250, but this is okay, I have nitpicked a lot on this PR :P